### PR TITLE
Prevent long chats from exceeding model limits

### PR DESCRIPTION
**Pull request by sachaservan**
## Summary
- budget system prompts, tools, attachments, output headroom, and complete conversation turns
- use the smallest eligible Auto model context window across request, archive, and usage indicators
- surface oversized requests through structured context-limit errors and safer attachment admission

## Testing
- `npx vitest run` (113 files, 1349 tests passed)
- `npx tsc --noEmit`
- targeted ESLint and Prettier checks

---
## Summary by cubic
Prevents long chats and large attachments from exceeding model limits by budgeting the full request (tokens and message count) and selecting only the newest complete turns that fit. Aligns archiving, usage, and attachment admission with the same live plan, and returns a structured `CONTEXT_LIMIT` when the newest turn can’t fit.

- **New Features**
  - Request planner budgets system instructions, GenUI tool schemas, time reminder, multimodal image allowance, router-tool headroom, tool-call overhead, and output reserve; keeps whole turns (incl. tool results).
  - Enforces server max-messages-per-request when provided; newest turn that exceeds tokens or message cap yields `CONTEXT_LIMIT` and the banner disables retry.
  - Uses the smallest eligible Auto context window across request building, archiving, and the usage bar; UI passes a computed `contextStartIndex` so both views match.
  - Attachment gate runs the live request plan (with current prompt, tools, and pending files) and rejects overflow with a clear toast.

- **Refactors**
  - Reworked budgeting into `planRequestContext`/`selectMessagesForRequest`, removed speculative budgeting, added per-message overhead, conservative image allowances, router tool headroom, and fixed message-limit handling.
  - `chat-query-builder` plans context with system/time/tool inputs and maps client/provider overflows to `CONTEXT_LIMIT`; tests cover edge cases and message-cap enforcement.

<sup>Written for commit 4e8841b2a8317d23c9e4b7f9d3194ab211f0e90c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/tinfoilsh/tinfoil-webapp/pull/464?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/components/chat/stream-error-banner.tsx [hunk 1]
```diff
@@ -173,10 +174,13 @@ export function StreamErrorBanner({
   // connection failures instead of the ambiguous "Try again" (which reads
   // like it might just reconnect). While offline the resend is gated —
   // it would only burn the automatic in-request retries and fail again.
-  const isConnectionError = kind === 'connection'
+  const isConnectionError = error.code === 'FETCH_ERROR'
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/stream-error-banner.tsx]: Retry behavior must use structured codes. Deriving it from kind would let message-string heuristics drive control flow, which is explicitly prohibited by AGENTS.md; the heuristic remains display-only.

File: src/utils/token-estimation.ts [hunk 2]
```diff
@@ -1,89 +1,334 @@
...
-    tokens += estimateTokenCount(msg.documentContent)
+  if (msg.documentContent) tokens += estimateTokenCount(msg.documentContent)
+  return tokens
+}
+
+function estimateUserContentTokens(
+  msg: Message,
+  isMultimodal: boolean,
+): number {
+  let textContent = msg.content
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/utils/token-estimation.ts]: The estimator is intentionally approximate, especially for image costs. A byte-identical formatting abstraction would add substantial coupling without fixing a current mismatch.

File: src/services/inference/chat-query-builder.ts [hunk 3]
```diff
@@ -212,6 +257,16 @@ export class ChatQueryBuilder {
     return content.trim() ? content : null
   }
 
+  private static buildPrependedSystemContent(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/services/inference/chat-query-builder.ts]: The separator difference is intentional for system-role versus prepended prompts. Keeping the two small helpers explicit is clearer than parameterizing a formatting detail.

File: src/components/chat/chat-messages.tsx [hunk 4]
```diff
@@ -336,12 +338,13 @@ export function ChatMessages({
   // Separate messages into archived and live sections - memoize this calculation
   const { archivedMessages, liveMessages } = useMemo(() => {
     const budget = getContextTokenBudget(currentModel?.contextWindow)
-    const startIndex = findContextStartIndex(messages, budget)
+    const startIndex =
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/chat-messages.tsx]: The fallback supports isolated component tests and non-production callers. It only affects display archiving, not the request payload, so making the prop required would add test churn without a correctness benefit.

File: src/components/chat/chat-interface.tsx [hunk 5]
```diff
@@ -2032,6 +2158,33 @@ export function ChatInterface({
...
+    quote,
+    selectedModel,
+    models,
+    systemPrompt: finalSystemPrompt,
+    rules: processedRules,
+    webSearchEnabled: effectiveWebSearchEnabled,
+    codeExecutionEnabled,
+    genUIEnabled,
+  })
+  requestContextRef.current = {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/chat-interface.tsx]: This is minor initialization boilerplate with no behavioral defect, and inference keeps both object shapes aligned. Refactoring it is not worth expanding this fix.

File: src/components/chat/chat-interface.tsx [hunk 6]
```diff
@@ -239,6 +242,129 @@ function buildAttachment(opts: {
   return undefined
 }
 
+function getCompletedAttachments(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/components/chat/chat-interface.tsx]: This helper also receives unfiltered state from context calculation and upload admission, so it must own the completion filters. The fallback preserves existing attachment submission semantics.

File: src/utils/token-estimation.ts [hunk 7]
```diff
@@ -1,89 +1,334 @@
+import {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply src/utils/token-estimation.ts]: attachment-helpers is a stateless data helper with type-only dependencies and no import back to token estimation, so there is no runtime cycle. Moving it would be architecture-only churn.

**Comment by sachaservan** [comment]
Closing after removing the speculative client-side request budgeting. Context enforcement and streaming failures should be fixed at the router/controlplane layer instead.
